src/communication/tlsvchannel.cpp

```diff
@@ -0,0 +1,210 @@
+    const aos::Array<uint8_t>& caCert, const aos::crypto::PrivateKeyItf& privKey, FuncDbg dbg)
+    : mCertChainPtr(certChain)
+    , mCACertArray(caCert)
+    , mPrivKey(privKey)
+    , mKeyID(0)
+    , mDbg(dbg)
+{
+}
+
+TLSVChannel::~TLSVChannel()
+{
+    mbedtls_x509_crt_free(&mCertChain);
+    mbedtls_x509_crt_free(&mCACert);
+    mbedtls_pk_free(&mPrivKeyCtx);
+    mbedtls_entropy_free(&mEntropy);
+    mbedtls_ctr_drbg_free(&mCtrDrbg);
+    mbedtls_ssl_config_free(&mConf);
+    mbedtls_ssl_free(&mSsl);
+
+    AosPsaRemoveKey(mKeyID);
+}
+
+aos::Error TLSVChannel::Init(const aos::String& xsReadPath, const aos::String xsWritePath)
+{
+    psa_status_t status = psa_crypto_init();
+    if (status != PSA_SUCCESS) {
+        return AOS_ERROR_WRAP(status);
+    }
+
+    const char* pers = "tls_vchannel_client";
+
+    mbedtls_x509_crt_init(&mCertChain);
+    mbedtls_x509_crt_init(&mCACert);
+    mbedtls_pk_init(&mPrivKeyCtx);
+    mbedtls_entropy_init(&mEntropy);
+    mbedtls_ctr_drbg_init(&mCtrDrbg);
+    mbedtls_ssl_config_init(&mConf);
+    mbedtls_ssl_init(&mSsl);
+
+    for (auto& cert : *mCertChainPtr) {
+        auto ret = mbedtls_x509_crt_parse(&mCertChain, cert.mRaw.Get(), cert.mRaw.Size());
+        if (ret != 0) {
+            return AOS_ERROR_WRAP(ret);
+        }
+    }
+
+    auto ret = mbedtls_x509_crt_parse(&mCACert, mCACertArray.Get(), mCACertArray.Size());
+
+    auto retKey = SetupOpaqueKey(mPrivKeyCtx, mPrivKey);
+    if (!retKey.mError.IsNone()) {
+        return retKey.mError;
+    }
+
+    mKeyID = retKey.mValue;
+
+    if ((ret = mbedtls_ctr_drbg_seed(
+             &mCtrDrbg, mbedtls_entropy_func, &mEntropy, (const unsigned char*)pers, strlen(pers)))
+        != 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    if ((ret = mbedtls_ssl_config_defaults(
+             &mConf, MBEDTLS_SSL_IS_CLIENT, MBEDTLS_SSL_TRANSPORT_STREAM, MBEDTLS_SSL_PRESET_DEFAULT))
+        != 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    mbedtls_ssl_conf_authmode(&mConf, MBEDTLS_SSL_VERIFY_REQUIRED);
+    mbedtls_ssl_conf_rng(&mConf, mbedtls_ctr_drbg_random, &mCtrDrbg);
+
+    mbedtls_ssl_conf_ca_chain(&mConf, &mCACert, nullptr);
+
+    if ((ret = mbedtls_ssl_conf_own_cert(&mConf, &mCertChain, &mPrivKeyCtx)) != 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    if ((ret = mbedtls_ssl_setup(&mSsl, &mConf)) != 0) {
+        return AOS_ERROR_WRAP(ret);
+    }
+
+    mbedtls_ssl_set_bio(&mSsl, this, TLSVChannel::TlsWrite, TLSVChannel::TlsRead, nullptr);
+
+#if defined(MBEDTLS_DEBUG_C)
+    mbedtls_debug_set_threshold(cDebugLevel);
+#endif
+
+    mbedtls_ssl_conf_dbg(&mConf, mDbg, nullptr);
+
+    return mVChannel.Init(xsReadPath, xsWritePath);
+}
+
+aos::Error TLSVChannel::Connect()
+{
+    mbedtls_ssl_session_reset(&mSsl);
+
+    auto err = mVChannel.Connect();
+    if (!err.IsNone()) {
+        return AOS_ERROR_WRAP(err);
+    }
+
+    int ret;
+    while ((ret = mbedtls_ssl_handshake(&mSsl)) != 0) {
```

Comment:
Do not wait, return err here. Communication will handle reconnect.

Reply:
Done. But in most examples doing like that. But ok let's return err